src/main/python/mmlspark/cyber/anomaly/collaborative_filtering.py

```diff
@@ -0,0 +1,880 @@
+                u) else np.array(v)
+            uu = np.pad(np.array(u), (0, len(v) - len(u)), 'constant', constant_values=1.0) if len(u) < len(
+                v) else np.array(u)
+
+            return float(vv.dot(uu))
+        else:
+            return None
+
+    return dot
+
+
+class CfAlgoParams:
+    """
+    Parameter for the AccessAnomaly Estimator indicating
+    if to use implicit or explicit feedback versions of the ALS algorithm
+    and additional parameters that are relevant for the implicit/explicit versions
+    """
+    def __init__(self, implicit: bool):
+        self._implicit = implicit
+        self._alpha: Optional[float] = None
+        self._complementset_factor: Optional[int] = None
+        self._neg_score: Optional[float] = None
+
+        # set default values
+        if implicit:
+            self.set_alpha(1.0)
+        else:
+            self.set_complementset_factor(2)
+            self.set_neg_score(1.0)
+
+    @property
+    def implicit(self) -> bool:
+        return self._implicit
+
+    @property
+    def alpha(self) -> Optional[float]:
+        """
+        Relevant for the implicit version only
+        :return: alpha value (see descriptions in explainParam of AccessAnomaly)
+        """
+        return self._alpha
+
+    @property
+    def complementset_factor(self) -> Optional[int]:
+        """
+        Relevant for the explicit version only
+        :return: complementset_factor value (see descriptions in explainParam of AccessAnomaly)
+        """
+        return self._complementset_factor
+
+    @property
+    def neg_score(self) -> Optional[float]:
+        """
+        Relevant for the explicit version only
+        :return: neg_score value (see descriptions in explainParam of AccessAnomaly)
+        """
+        return self._neg_score
+
+    def set_alpha(self, value: float):
+        assert self.implicit
+        self._alpha = value
+        return self
+
+    def set_complementset_factor(self, value: int):
+        assert not self.implicit
+        self._complementset_factor = value
+        return self
+
+    def set_neg_score(self, value: float):
+        assert not self.implicit
+        assert value > 0.0
+        self._neg_score = value
+        return self
+
+
+class AccessAnomalyConfig:
+    """
+    Define default values for AccessAnomaly Params
+    """
+    default_tenant_col = 'tenant'
+    default_user_col = 'user'
+    default_res_col = 'res'
+    default_likelihood_col = 'likelihood'
+    default_output_col = 'anomaly_score'
+
+    default_rank = 10
+    default_max_iter = 25
+    default_reg_param = 1.0
+    default_num_blocks = None   # |tenants| if separate_tenants is False else 10
+    default_separate_tenants = False
+
+    default_low_value = 5.0
+    default_high_value = 10.0
+
+    default_algo_cf_params = CfAlgoParams(True)
+
+
+class UserResourceFeatureVectorMapping:
```

Comment:
This shouldnt be public but rather something that happens during estimator fitting

Reply:
Made this class private as suggested

Reply:
done